mautic/mautic
# Separate assets from media

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
**why**

make the lifecycle of composer based installs easier by not mixing application content and user uploaded content in the same folders.
This PR created a distinction between `app/assets` (application)  and ` media` (user) files.

**what does this PR**
* move all the static assets from the `media` folder to the `app/assets` folder
    * the dasboards
    * the flags
    * the mejs files
* introduce the `AssetsHelper::getOverridableUrl()` method, to provide logic to override an Mautic asset in `app/assets` via a file in the `media` folder
    * use the method above in all relevant places where logo's, flags, ... are references
* a command `mautic:assets:cleanup` that cleans up files in the `media` folder that already exist in the `app/assets` folder, and are exactly the same based on the md5 hash.

**possible folluw-up tasks**
* use the term `asset` and `media` consistently in code


#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. check for broken images / js / css files in the installer, when clicking around, or when hitting an error page
3. test the asset override functionality
    - add a contact and assign a country (e.g. Belgium)
    - check if the flag loads
    - override the flag by copying another flag from the `app/assets` folder to the `media` folder
      ```
      cp app/assets/images/flags/France.png media/images/flags/Belgium.png
      ```
    - refresh the contact info and verify the flag is overridden.
4. test the `mautic:assets:cleanup` command
    - copy an extra asset to the media folder with the 
      ```
      cp app/assets/images/mautic_logo_db64.png media/images/mautic_logo_db64.png
      ```
    - run the cleanup command, it should show 1 file
      ```
      ./bin/console mautic:assets:cleanup
      1 obsolete files found
       - /var/www/html/mautic/media/images/mautic_logo_db64.png
      ```
    - verify `media/images/mautic_logo_db64.png` is gone, but `media/images/flags/Belgium.png` still exists



<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10682"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
